File: app/src/main/java/com/bannersimple/simple/RefreshActivity.java
package com.bannersimple.simple;

import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.widget.SwipeRefreshLayout;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.widget.Toast;

import com.bannersimple.R;
import com.bannersimple.bean.SimpleData;
import com.bannersimple.refresh.Api;
import com.bannersimple.refresh.ListModel;
import com.bannersimple.refresh.RefreshAdapter;

import java.util.ArrayList;
import java.util.List;

import io.reactivex.network.manager.RxNetWork;
import io.reactivex.network.manager.RxNetWorkListener;

/**
 * by y on 2017/3/8.
 */

public class RefreshActivity extends AppCompatActivity
        implements SwipeRefreshLayout.OnRefreshListener,
        RxNetWorkListener<List<ListModel>>, Runnable {


    private SwipeRefreshLayout swipeRefreshLayout;
    private RefreshAdapter adapter;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_refresh);
        swipeRefreshLayout = (SwipeRefreshLayout) findViewById(R.id.refresh_layout);
        RecyclerView recyclerView = (RecyclerView) findViewById(R.id.recyclerView);

        swipeRefreshLayout.setOnRefreshListener(this);
        swipeRefreshLayout.post(this);

        adapter = new RefreshAdapter(new ArrayList<ListModel>());
        recyclerView.setLayoutManager(new LinearLayoutManager(getApplicationContext()));
        recyclerView.setAdapter(adapter);
    }

    @Override
    public void onRefresh() {
        RxNetWork.getInstance().cancel(RxNetWork.TAG);
        RxNetWork
                .getInstance()
                .setBaseUrl(Api.ZL_BASE_API)
                .getApi(RxNetWork.TAG,
                        RxNetWork.observable(Api.ZLService.class).getList("daily", 20, 0), this);
    }


    @Override
    public void onNetWorkStart() {
        swipeRefreshLayout.setRefreshing(true);
    }

    @Override
    public void onNetWorkError(Throwable e) {
        swipeRefreshLayout.setRefreshing(false);
        Toast.makeText(getApplicationContext(), "net work error", Toast.LENGTH_SHORT).show();
    }

    @Override
    public void onNetWorkComplete() {
        swipeRefreshLayout.setRefreshing(false);
    }

    @Override
    public void onNetWorkSuccess(List<ListModel> data) {
        adapter.clear();
        adapter.addBanner(SimpleData.initModel());
        adapter.addAll(data);
    }


    @Override
    protected void onDestroy() {
        super.onDestroy();
        adapter.clearBanner();
        RxNetWork.getInstance().cancel(RxNetWork.TAG);
    }

    @Override
    public void run() {
        onRefresh();
    }
}

File: app/src/main/java/com/bannersimple/simple/RefreshActivity.kt
package com.bannersimple.simple

import android.os.Bundle
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout
import com.bannersimple.R
import com.bannersimple.bean.SimpleData
import com.bannersimple.refresh.Api
import com.bannersimple.refresh.ListModel
import com.bannersimple.refresh.RefreshAdapter
import io.reactivex.network.RxNetWork
import io.reactivex.network.RxNetWorkListener
import java.util.*


/**
 * by y on 2017/3/8.
 */

class RefreshActivity : AppCompatActivity(), SwipeRefreshLayout.OnRefreshListener, RxNetWorkListener<ListModel>, Runnable {

    private lateinit var swipeRefreshLayout: SwipeRefreshLayout
    private lateinit var adapter: RefreshAdapter

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_refresh)
        swipeRefreshLayout = findViewById(R.id.refresh_layout)
        val recyclerView = findViewById<RecyclerView>(R.id.recyclerView)
        swipeRefreshLayout.setOnRefreshListener(this)
        swipeRefreshLayout.post(this)
        adapter = RefreshAdapter(ArrayList())
        recyclerView.layoutManager = LinearLayoutManager(applicationContext)
        recyclerView.adapter = adapter
    }

    override fun onRefresh() {
        RxNetWork.instance.cancel(javaClass.simpleName)
        RxNetWork
                .instance
                .getApi(javaClass.simpleName, RxNetWork.observable(Api.ZLService::class.java).getList(), this)
    }


    override fun onNetWorkStart() {
        swipeRefreshLayout.isRefreshing = true
    }

    override fun onNetWorkError(e: Throwable) {
        swipeRefreshLayout.isRefreshing = false
        Toast.makeText(applicationContext, e.message, Toast.LENGTH_SHORT).show()
    }

    override fun onNetWorkComplete() {
        swipeRefreshLayout.isRefreshing = false
    }

    override fun onNetWorkSuccess(data: ListModel) {
        adapter.clear()
        adapter.addBanner(SimpleData.initModel())
        adapter.addAll(data.top_stories)
    }


    override fun onDestroy() {
        super.onDestroy()
        RxNetWork.instance.cancel(javaClass.simpleName)
    }

    override fun run() {
        onRefresh()
    }
}
